File: 287-find-duplicate-number/go/main.go
package main

func findDuplicate(nums []int) int {
	s := 0
	f := 0
	s = nums[s]
	f = nums[nums[f]]

	for s != f {
		s = nums[s]
		f = nums[nums[f]]
	}

	s = 0

	for s != f {
		s = nums[s]
		f = nums[f]
	}

	return f
}

File: 287-find-duplicate-number/cpp/main.cpp
class Solution {
 public:
  int findDuplicate(vector<int>& nums) {
    int s = 0;
    int f = 0;
    s = nums[s];
    f = nums[nums[f]];

    while (f != s) {
      s = nums[s];
      f = nums[nums[f]];
    }

    s = 0;
    while (f != s) {
      s = nums[s];
      f = nums[f];
    }

    return f;
  }
};
